lib/matplotlib/figure.py

```diff
@@ -1462,67 +1518,22 @@ def legend(self, *args, **kwargs):
         :ref:`sphx_glr_tutorials_intermediate_legend_guide.py` for details.
         """
 
-        # If no arguments given, collect up all the artists on the figure
-        if len(args) == 0:
-            handles = []
-            labels = []
-
-            def in_handles(h, l):
-                # Method to check if we already have a given handle and label.
-                # Consider two handles to be the same if they share a label,
-                # color, facecolor, and edgecolor.
-
-                # Loop through each handle and label already collected
-                for f_h, f_l in zip(handles, labels):
-                    if f_l != l:
-                        continue
-                    if type(f_h) != type(h):
-                        continue
-                    try:
-                        if f_h.get_color() != h.get_color():
-                            continue
-                    except AttributeError:
-                        pass
-                    try:
-                        if f_h.get_facecolor() != h.get_facecolor():
-                            continue
-                    except AttributeError:
-                        pass
-                    try:
-                        if f_h.get_edgecolor() != h.get_edgecolor():
-                            continue
-                    except AttributeError:
-                        pass
-                    return True
-                return False
-
-            for ax in self.axes:
-                ax_handles, ax_labels = ax.get_legend_handles_labels()
-                for h, l in zip(ax_handles, ax_labels):
-                    if not in_handles(h, l):
-                        handles.append(h)
-                        labels.append(l)
-            if len(handles) == 0:
-                warnings.warn("No labeled objects found. "
-                              "Use label='...' kwarg on individual plots.")
-                return None
-
-        elif len(args) == 2:
-            # LINES, LABELS
-            handles, labels = args
-
-        elif len(args) == 3:
-            # LINES, LABELS, LOC
-            handles, labels, loc = args
-            kwargs['loc'] = loc
-
-        else:
-            raise TypeError('Invalid number of arguments passed to legend. '
-                            'Please specify either 0 args, 2 args '
-                            '(artist handles, figure labels) or 3 args '
-                            '(artist handles, figure labels, legend location)')
-
-        l = Legend(self, handles, labels, **kwargs)
+        handles, labels, extra_args = mlegend._parse_legend_args(self.axes,
+                *args, **kwargs)
+        # these are popped in the above...
+        kwargs.pop('handles', None)
+        kwargs.pop('labels', None)
+        # check for third arg
+        if len(extra_args):
+            cbook.warn_deprecated(
```

Comment:
It is too late to deprecate this for 2.1 if we take semver seriousishly.

Reply:
Oops, sorry.  What should go there?  `2.1.1`?  `2.2`?

Reply:
What @tacaswell means is this PR can't be backported if the deprecation is made (or at least the deprecation part should be a separate PR that isn't backported.)

Reply:
OK, so remove the deprecation for now, and then when its merged make a new PR?   Thats easy enough....

Reply:
Thanks, sorry for being pesky about this.

Reply:
No problem at all.  I'm new here, and don't have any experience with the versioning in or outs.

Reply:
Deprecations are [point 7 in semver](http://semver.org/#spec-item-7); you can read the rest at your leisure.